Move to eio

clecat opened the pull request:
This pull requests goal is to finalize to move from lwt to eio within the irmin project, it should be merged and reviewed once we are satisfied by the state of the eio branch and most of the PRs pointing to it are merged.

It will also allow to keep an eye on the status of the CI.

patricoferris committed:
Convert Irmin from Lwt to direct-style with Eio

patricoferris committed:
Restore Test_tree.clear tree size

patricoferris committed:
Fix map_p, filter_map_p and convert irmin-fs

patricoferris committed:
Add eio to opam file

patricoferris committed:
Convert irmin-pack to direct-style

patricoferris committed:
Update to latest eio

patricoferris committed:
Working FS unix backend

patricoferris committed:
Convert irmin-containers to direct-style

patricoferris committed:
Remove pin-depends for eio

patricoferris committed:
Fix irmin-fs file info

patricoferris committed:
Fix ignoring commit in store

patricoferris committed:
Fix generic-key tests

patricoferris committed:
Remove traceln

patricoferris committed:
Update to eio.0.7 and catch IO errors

ElectreAAS committed:
Add lock on branch

art-w committed:
Fix after rebase

art-w committed:
Fix irmin-pack bench

art-w committed:
Fix irmin-pack examples

art-w committed:
Fix irmin-http using lwt_eio

art-w committed:
Fix irmin-graphql using lwt_eio

art-w committed:
Fix irmin-git using lwt_eio

art-w committed:
Fix irmin-cli using lwt_eio

art-w committed:
Fix irmin-mirage using lwt_eio

art-w committed:
Fix examples using lwt_eio

art-w committed:
Fix for libirmin

ElectreAAS committed:
Add multicore tests

ElectreAAS committed:
Fix LRU hashtable with a mutex

ElectreAAS committed:
Fixed data race in stat counting

ElectreAAS committed:
Fixed data race in tree counters

ElectreAAS committed:
Fixed again stats, this time not in unix

ElectreAAS committed:
Update multicore testing

ElectreAAS committed:
Added locks on unix/pack-store, in Tbl

ElectreAAS committed:
Added locks in irmin-pack atomic_write

ElectreAAS committed:
Added locks in control-file

ElectreAAS committed:
Remove unsafe index stats in irmin-pack io

ElectreAAS committed:
Added locks in store

art-w committed:
Safe pack_key promotion

art-w committed:
Optimize control file payload

art-w committed:
Remove unsafe usage of Index.Stats

art-w committed:
Fix Tree findv_cache

art-w committed:
Fix Tree value cache

art-w committed:
Remove spinlock on Tree stats

art-w committed:
Fix irmin-pack inode partial_ptr

art-w committed:
Fix opam dependencies to lwt_eio

art-w committed:
irmin-pack: add multicore unit tests

art-w committed:
Fix alcotest dependency

art-w committed:
Fix irmin Tree lazy length

art-w committed:
Fix irmin Tree Contents mutables

art-w committed:
Fix irmin proof

art-w committed:
Fix irmin Tree add and remove

art-w committed:
Fix irmin-pack Append_only_file buffer length

art-w committed:
Fix irmin-pack Gc cancellation on close

art-w committed:
Fix irmin-pack Async Gc status handling

art-w committed:
Fix for append_only_file buffer length

art-w committed:
Fix index-unix flushing from another thread

art-w committed:
Fix irmin-pack Gc locks

art-w committed:
Add commit unit test

art-w committed:
Fix irmin-pack batch in batch test

art-w committed:
Fix irmin store mutable tree

art-w committed:
Fix commit write lock

art-w committed:
Fix parallel tree hash computation

clecat committed:
irmin-pack: Add Store.Tree.list test

art-w committed:
Fix libirmin: at_exit executes outside of eio

clecat committed:
irmin-pack: Add tests on commits for of_hash & parents

clecat committed:
irmin-pack: Add Commit.v test & fix it with a lock

art-w committed:
Add half and full diamond multicore benchmarks

art-w committed:
Optimize Tree findv_cache for add/remove

art-w committed:
Fix formatting and remove old tests

art-w committed:
Fix irmin_fsck

art-w committed:
Fix opam dependencies

art-w committed:
Fix opam dependencies

art-w committed:
irmin-pack: remove Io.Unix constraint

art-w committed:
eio: irmin-server using lwt_eio

art-w committed:
upgrade to eio.0.12

art-w committed:
irmin-pack: generic io lib without unix dependency

art-w committed:
irmin-pack: generic io lib without unix dependency

art-w committed:
irmin-pack: extract unix Async

art-w committed:
irmin-pack: extract remaining Unix functions

art-w committed:
irmin-pack: extract remaining Unix functions

art-w committed:
irmin-pack: fix Gc.on_finalise

art-w committed:
mirage: irmin-pack.unix is optional

art-w committed:
irmin-pack.io: remove rusage dependency

art-w committed:
mirage: remove dependency to fmt_tty

art-w committed:
mirage: remove dependency to mtime.clock.os

art-w committed:
mirage: remove dependency to progress

art-w committed:
mirage: remove dependency to progress

art-w committed:
fix progress dependency injection

art-w committed:
Update opam dependencies

art-w committed:
Update opam dependencies

art-w committed:
irmin-pack: use checkseum.c

art-w committed:
Fix irmin-client tests

clecat committed:
irmin: Move LRU cache to kcas

clecat committed:
irmin-pack: Change hash tables to kcas

clecat committed:
irmin: Bring iter back

art-w committed:
Simplify concurrent list operations

art-w committed:
Fix CI opam dependencies

art-w committed:
Fix CI opam dependencies

art-w committed:
Upgrade dune and fix new warnings

art-w committed:
macos: reduce number of open files in unix-client test to avoid EMFILE error

art-w committed:
Upgrade to Eio 1.0

art-w committed:
Update CHANGES

art-w committed:
Temporarily disable CI for specific platforms issues

art-w committed:
Temporarily disable CI for specific platforms issues

art-w committed:
Update opam for OCaml 5.1.0

art-w committed:
irmin-fs: eio backend

clecat committed:
irmin-pack: eio backend

art-w committed:
libirmin: reuse a single eio scheduler across calls

art-w committed:
Update CI for OCaml 5.2

art-w committed:
Fix multicore race in unit test

clecat committed:
Remove pin-depend to alcotest

clecat committed:
Apply fmt

clecat committed:
Fix irmin-client test

lyrm committed:
Remove pin-depends to progress and terminal and remove irmin-http.opam file.

lyrm committed:
Remove useless lwt dependencies.

clecat committed:
Remove _lwt from function names

clecat committed:
Merge pull request #2149 from patricoferris/direct-style

[RFC]: Convert to direct-style with Eio

clecat committed:
Remove unnecessary destructive substitution

clecat committed:
Rename key'

clecat committed:
Merge pull request #2316 from art-w/eio-fs-conf

irmin-fs: Eio backend

clecat committed:
Fix the CI

clecat committed:
Fix the CI

clecat committed:
Fix CI

clecat committed:
Add logs threaded safety to tests

clecat committed:
Up dockerfile ocaml version

clecat committed:
Use dls in irmin stats

clecat committed:
Give Alcotest self defined stdout/stderr through irmin-test

clecat committed:
Remove unnecessary TODOs, clean types

lyrm committed:
Add qcheck-stm test for LRU cache.

clecat committed:
Fix lint

clecat committed:
Clean fs & sw

lyrm committed:
Add doc in lru.mli and improve the add function.

lyrm committed:
Move lru cache weight management from irmin-pack to irmin. This enables to ensure multicore safety for both lru caches.

samoht committed:
Merge pull request #2280 from clecat/move-io-to-eio

Move Irmin_pack.Io to eio

samoht committed:
Merge pull request #2281 from clecat/kcas-lru

Change LRU cache and other hash tables to use kcas

cuihtlauac committed:
Fix reverse Fisher-Yates shuffle

lyrm commented:
I made a list of things that needed fixes [here ](https://github.com/mirage/irmin/pull/2149#pullrequestreview-3029765907). This is clearly outdated, but it's likely that not everything has been done.  A lot of it is small tasks (updating examples and documentation), and it was decided that it would be done before merging to main, if I remembered correctly.

zshipko committed:
Merge pull request #2377 from cuihtlauac/fix-shuffle

Fix array shuffle

cuihtlauac committed:
Fix libirmin build

zshipko committed:
Merge pull request #2380 from cuihtlauac/remove-libirmin-warnings

Fix libirmin build

zshipko committed:
re-enable on macos

zshipko committed:
fix libirmin tests on linux

zshipko committed:
format discover.ml

zshipko committed:
prefer enabled_if in dune file to discover.ml script

zshipko committed:
extract libirmin foreign stubs into a separate library stanza

zshipko committed:
try requiring newer dune to avoid crash in lower bounds tests

zshipko committed:
Merge pull request #2318 from art-w/eio-libirmin

libirmin: reuse a single eio scheduler across calls

zshipko committed:
add merkle_proofs.md to irmin-cli package

zshipko committed:
add README.md to irmin-cli package

zshipko committed:
fix deprecated qcheck functions

zshipko committed:
Merge pull request #2384 from mirage/eio-setup

Fix repo initialization with opam

zshipko committed:
Improve graphql docs (#2383)

zshipko committed:
improve irmin-fs path handling (#2382)

lyrm committed:
Multicore safety tests with qcheck-stm  (#2373)

* Qcheck stm test for irmin

* Add irmin-pack tests

* Some debugging.

* Fix dune file.

* Replace deprecated functions

* Update pin to qcheck-stm

zshipko committed:
eio: sync with main (#2385)

* Add '/' to allowed branch characters

Signed-off-by: Benjamin Somers <benjamin.somers@imt-atlantique.fr>

* Add funding manifest discovery

See https://floss.fund/funding-manifest/

* Update manifest URL

---------

Signed-off-by: Benjamin Somers <benjamin.somers@imt-atlantique.fr>
Co-authored-by: Benjamin Somers <benjamin.somers@imt-atlantique.fr>
Co-authored-by: Thomas Gazagnaire <thomas@gazagnaire.org>

cuihtlauac committed:
Fix test_stm by using posix Eio backend (#2391)

* Fix test_stm by using posix Eio backend

The test_stm test was disabled because it exhausted io_uring resources
on Linux (ENOMEM on io_uring_queue_init when spawning multiple Eio
domains).

Instead of disabling the test, use the posix backend via EIO_BACKEND
environment variable. This requires converting the test stanza to an
executable + rule pattern to allow setting the environment variable.

ElectreAAS committed:
Fix formatting in test logs (#2387)

* Fix formatting in test logs

Signed-off-by: Ambre Austen Suhamy <ambre@tarides.com>

* Update ocamlformat 0.28.1

* ppx_irmin: use ppxlib >= 0.37

* pin bisect_ppx

---------

Signed-off-by: Ambre Austen Suhamy <ambre@tarides.com>
Co-authored-by: Arthur Wendling <arthur@tarides.com>
Co-authored-by: zach <zachshipko@gmail.com>

ElectreAAS committed:
Add CI job building irmin with dune package management (#2388)

* Fix irmin-watcher hash to newest

Signed-off-by: Ambre Austen Suhamy <ambre@tarides.com>

* Added setup-dune CI workflow

Signed-off-by: Ambre Austen Suhamy <ambre@tarides.com>

* fix formatting

---------

Signed-off-by: Ambre Austen Suhamy <ambre@tarides.com>
Co-authored-by: zach <zachshipko@gmail.com>

cuihtlauac committed:
Fix duplicate pin-depends (#2395)

balat committed:
Bump ocamlformat to 0.29.0 and reformat
